Merge guidelines from RFC 1567 into UX Guidelines.

Havvy opened the pull request:
This is a partial fix for issue #34808.

Most of the wording was copied verbatim from the RFC. Main difference is that I moved the actual template to the top of the section.

It also makes the error explanations the longest section in the guidelines doc for now.

rust-highfive commented:
r? @steveklabnik

(rust_highfive has picked a reviewer for you, use r? to override)

steveklabnik commented:
This looks decent to me but I don't have time to do a full review right now, if anyone else feels like it's good, feel free to r+

steveklabnik commented:
@bors: r+ rollup

Thanks! :)

bors commented:
:pushpin: Commit 010e479 has been approved by `steveklabnik`

<!-- @bors r=steveklabnik 010e479fa270c8e93237cbacbef8fec37f57cf73 -->

bors commented:
:hourglass: Testing commit 010e479 with merge cda0675...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5088)

alexcrichton commented:
@bors: retry

On Tue, Jul 26, 2016 at 4:16 AM, bors <notifications@github.com> wrote:

> 💔 Test failed - auto-win-msvc-64-opt
> <https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5088>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34894#issuecomment-235239212>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95Csr5A8xmi0Bn-fQ5N36aIPJZuc5ks5qZeyjgaJpZM4JOWxQ>
> .
>

steveklabnik commented:
@bors: r-

I _believe_ that this caused https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/1986

steveklabnik commented on src/doc/rustc-ux-guidelines.md:
````diff
@@ -74,6 +78,124 @@ code with backticks.
+\```
+
+[Optional Additional information]
+```
````
looks like this wasn't closed here

steveklabnik commented:
@bors: r+ rollup

bors commented:
:pushpin: Commit f1a51bd has been approved by `steveklabnik`

<!-- @bors r=steveklabnik f1a51bd41a4438da0532e8b6ea29bff479a7a90e -->

GuillaumeGomez commented:
@bors: r-

It definitely makes windows build fail. You have to add ignore flag on every code block.

Havvy commented:
Actually, this shows a bigger issue. I want to show the example and the template as raw text, not as markdown. But I don't know how to do that...I was hoping the blockquotes did it, but it doesn't. It looks like the backslash from the template also gets put into the output.

steveklabnik commented:
use 

    ```text

Havvy commented:
Using ```text doesn't work here, because it stops at the first lone ``` it encounters.

Havvy commented:
I think this will work. Rebased my messing around with markdown into one commit. Also fixed up some of the attempts of escaping markdown from the original template in the RFC.

I hope everything is good now?

GuillaumeGomez commented:
Seems good for me.

cc @steveklabnik

steveklabnik commented:
@bors: r+ rollup

bors commented:
:pushpin: Commit cede35e has been approved by `steveklabnik`

<!-- @bors r=steveklabnik cede35eaada00e4993e5ae4400689541837d445d -->

GuillaumeGomez commented:
It still doesn't work:

```
doc tests for: C:\bot\slave\auto-win-gnu-32-opt-rustbuild\build\src/doc\rustc-ux-guidelines.md

running 6 tests
test _1 ... ignored
test _2 ... ignored
test _3 ... ignored
test _4 ... ignored
test _5 ... FAILED
test _0 ... FAILED

failures:

---- _5 stdout ----
	<anon>:2:12: 41:4 error: expected type, found `r##"
An "or" pattern was used where the variable bindings are not consistently bound
across patterns.

Example of erroneous code:

```compile_fail,E0409
let x = (0, 2);
match x {
    (0, ref y) | (y, 0) ={ /* use y */} // error: variable `y` is bound with
                                          //        different mode in pattern #2
                                          //        than in pattern #1
    _ =()
}
```

Here, `y` is bound by-value in one case and by-reference in the other.

To fix this error, just use the same mode in both cases.
Generally using `ref` or `ref mut` where not already used will fix this:

```rust
let x = (0, 2);
match x {
    (0, ref y) | (ref y, 0) ={ /* use y */}
    _ =()
}
```

Alternatively, split the pattern:

```rust
let x = (0, 2);
match x {
    (y, 0) ={ /* use y */ }
    (0, ref y) ={ /* use y */}
    _ =()
}
```
"##` 
<anon>:2     E0409: r##"
                    ^
error: aborting due to previous error(s) 
thread '_5' panicked at 'Box<Any>', C:\bot\slave\auto-win-gnu-32-opt-rustbuild\build\src\librustc\session/mod.rs:168

---- _0 stdout ----
	<anon>:2:11: 19:4 error: expected type, found `r##"
[Error description]

Example of erroneous code:

```compile_fail
[Minimal example]
```

[Error explanation]

```
[How to fix the problem]
```

[Optional Additional information]

"##` 
<anon>:2     E000: r##"
                   ^
error: aborting due to previous error(s) 
thread '_0' panicked at 'Box<Any>', C:\bot\slave\auto-win-gnu-32-opt-rustbuild\build\src\librustc\session/mod.rs:168
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    _0
    _5

test result: FAILED. 0 passed; 2 failed; 4 ignored; 0 measured


command did not execute successfully: "C:\\bot\\slave\\auto-win-gnu-32-opt-rustbuild\\build\\obj\\build\\i686-pc-windows-gnu\\stage2\\bin\\rustdoc.exe" "--test" "C:\\bot\\slave\\auto-win-gnu-32-opt-rustbuild\\build\\src/doc\\rustc-ux-guidelines.md" "--test-args" ""
expected success, got: exit code: 101
```

GuillaumeGomez commented:
@bors: r-

steveklabnik commented:
@Havvy ping! Still keeping this PR going?

Havvy commented:
I'm stuck at the moment because I'm not sure how to show markdown that includes code blocks without putting ignore on them. I was going to see if I could mess with the test finder code to see what's going on. And then other duties took over while I was looking into it.

Havvy commented:
Closing this until I have more time to look at it.
